### PR TITLE
Migrate FlutterView, FlutterPlatformViews, FlutterOverlayView to ARC

### PR DESCRIPTION
**Pull request by jmagman**
Smart pointers support ARC as of https://github.com/flutter/engine/pull/47612, and the unit tests were migrated in https://github.com/flutter/engine/pull/48162.

Migrate `FlutterView`, `FlutterPlatformViews`, and `FlutterOverlayView` from MRC to ARC. 

Part of https://github.com/flutter/flutter/issues/137801.

### CODE REVIEW

File: shell/platform/darwin/ios/BUILD.gn [hunk 1]
```diff
@@ -87,6 +92,9 @@ source_set("flutter_framework_source_arc") {
     "framework/Source/FlutterUndoManagerDelegate.h",
     "framework/Source/FlutterUndoManagerPlugin.h",
     "framework/Source/FlutterUndoManagerPlugin.mm",
+    "framework/Source/FlutterView.h",
+    "framework/Source/FlutterView.mm",
+    "framework/Source/FlutterViewResponder.h",
```
Comment (jmagman) [inline shell/platform/darwin/ios/BUILD.gn:97]: `FlutterViewResponder.h` wasn't included anywhere?

File: shell/platform/darwin/ios/framework/Source/FlutterPlatformViews_Internal.h [hunk 2]
```diff
@@ -210,16 +210,19 @@ class FlutterPlatformViewsController {
 
   fml::WeakPtr<flutter::FlutterPlatformViewsController> GetWeakPtr();
 
-  void SetFlutterView(UIView* flutter_view);
+  void SetFlutterView(UIView* flutter_view) __attribute__((cf_audited_transfer));
```
Comment (jmagman) [inline shell/platform/darwin/ios/framework/Source/FlutterPlatformViews_Internal.h:213]: See `cf_audited_transfer` discussion https://github.com/flutter/engine/pull/52226#discussion_r1572924618

File: shell/platform/darwin/ios/framework/Source/SemanticsObject.mm [hunk 3]
```diff
@@ -867,7 +867,7 @@ - (UIAccessibilityTraits)accessibilityTraits {
 @end
 
 @interface FlutterPlatformViewSemanticsContainer ()
-@property(nonatomic, retain) UIView* platformView;
+@property(nonatomic, assign) UIView* platformView;
```
Comment (jmagman) [inline shell/platform/darwin/ios/framework/Source/SemanticsObject.mm:870]: See https://github.com/flutter/engine/pull/52484, platformView needs to be weak here.  I'll figure out a test for that one first and let that land.

File: testing/ios/IosUnitTests/IosUnitTests.xcodeproj/project.pbxproj [hunk 4]
```diff
@@ -99,6 +100,7 @@
 		0AC232E924BA71D300A85907 /* Source */ = {
 			isa = PBXGroup;
 			children = (
+				F76A3A892BE48F2F00A654F1 /* FlutterPlatformViewsTest.mm */,
```
Comment (jmagman) [inline testing/ios/IosUnitTests/IosUnitTests.xcodeproj/project.pbxproj:103]: Just a convenience to add it here, the test dylib made from BUILD.gn already built it.

File: shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm [hunk 5]
```diff
@@ -417,7 +416,7 @@ static bool ClipRRectContainsPlatformViewBoundingRect(const SkRRect& clip_rrect,
 
 long FlutterPlatformViewsController::FindFirstResponderPlatformViewId() {
   for (auto const& [id, root_view] : root_views_) {
-    if ((UIView*)(root_view.get()).flt_hasFirstResponderInViewHierarchySubtree) {
+    if (((UIView*)root_view.get()).flt_hasFirstResponderInViewHierarchySubtree) {
```
Comment (jmagman) [inline shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm:419]: `error: cast of 'BOOL' (aka 'bool') to 'UIView *' is disallowed with ARC`
```
../../../flutter/shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm:419:9: error: cast of 'BOOL' (aka 'bool') to 'UIView *' is disallowed with ARC
  419 |     if ((UIView*)(root_view.get()).flt_hasFirstResponderInViewHierarchySubtree) {
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
```
so... was this ever working?

File: shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm [hunk 6]
```diff
@@ -1060,19 +1057,12 @@ - (void)touchesEnded:(NSSet*)touches withEvent:(UIEvent*)event {
 }
 
 - (id)accessibilityContainer {
-  return _flutterAccessibilityContainer;
-}
-
-- (void)dealloc {
-  [_flutterAccessibilityContainer release];
-  [super dealloc];
+  return self.flutterAccessibilityContainer;
```
[the remark this answers is not in the document]
Comment (tacck) [reply shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm:1060]: I think it's LGTM!

File: shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm (hunk 5 above)
Comment (hellohuanlin) [reply shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm:419]: weird

**Review by hellohuanlin — APPROVED** [review]

**Pull request merged**